cli/printlog/printlog.go

```diff
@@ -46,10 +50,244 @@ func HandlePrint(args []string) (int, error) {
 		}
 		return 0, nil
 	}
+	if *compactExecLog != "" {
+		if err := printCompactExecLog(*compactExecLog); err != nil {
+			return -1, err
+		}
+		return 0, nil
+	}
 	log.Print(usage)
 	return 1, nil
 }
 
+func printCompactExecLog(path string) error {
+	f, err := os.Open(path)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+	r, err := zstd.NewReader(f)
+	if err != nil {
+		return err
+	}
+	defer r.Close()
+	slr := NewSpawnLogReconstructor(r)
+
+	for {
+		sr, err := slr.GetSpawnExec()
+		if err == io.EOF {
+			return nil
+		}
+		if err != nil {
+			return fmt.Errorf("failed getting spawn exec: %s", err)
+		}
+		b, err := protojson.MarshalOptions{Multiline: true}.Marshal(sr)
+		if err != nil {
+			return fmt.Errorf("failed to marshal remote gRPC log entry: %s", err)
+		}
+		if _, err := os.Stdout.Write(append(b, []byte{'\n'}...)); err != nil {
+			return fmt.Errorf("failed to write to stdout: %s", err)
+		}
+	}
+}
+
+type SpawnLogReconstructor struct {
+	input io.Reader
+
+	hashFunc string
+	files    map[int32]*spb.File
+	dirs     map[int32]*reconstructedDir
+	symlinks map[int32]*spb.File
+	sets     map[int32]*spb.ExecLogEntry_InputSet
+}
+
+type reconstructedDir struct {
+	path  string
+	files []*spb.File
+}
+
+func NewSpawnLogReconstructor(input io.Reader) *SpawnLogReconstructor {
+	return &SpawnLogReconstructor{
+		input:    input,
+		hashFunc: "",
+		files:    make(map[int32]*spb.File),
+		dirs:     make(map[int32]*reconstructedDir),
+		symlinks: make(map[int32]*spb.File),
+		sets:     make(map[int32]*spb.ExecLogEntry_InputSet),
+	}
+}
+
+func (slr *SpawnLogReconstructor) GetSpawnExec() (*spb.SpawnExec, error) {
+	entry := &spb.ExecLogEntry{}
+	br := bufio.NewReader(slr.input)
+	for {
+		err := protodelim.UnmarshalFrom(br, entry)
+		if err == io.EOF {
+			return nil, io.EOF
+		}
+		if err != nil {
+			return nil, fmt.Errorf("failed to read execution log entry: %s", err)
+		}
+
+		switch e := entry.GetType().(type) {
+		case *spb.ExecLogEntry_Invocation_:
+			slr.hashFunc = e.Invocation.GetHashFunctionName()
+		case *spb.ExecLogEntry_File_:
+			slr.files[entry.GetId()] = reconFile(nil, e.File)
+		case *spb.ExecLogEntry_Directory_:
+			slr.dirs[entry.GetId()] = reconDir(e.Directory)
+		case *spb.ExecLogEntry_UnresolvedSymlink_:
+			slr.symlinks[entry.GetId()] = reconSymlink(e.UnresolvedSymlink)
+		case *spb.ExecLogEntry_InputSet_:
+			slr.sets[entry.GetId()] = e.InputSet
+		case *spb.ExecLogEntry_Spawn_:
+			return slr.reconSpawn(e.Spawn)
+		default:
+			return nil, fmt.Errorf("unknown entry: %s", entry)
```

Comment:
for forwards-compatibility should we just log a warning here?

Reply:
Oh good call. I did not think of that.

Switched to `log.Warnf`